### ISSUE TITLE
Assistance solving a multi-mod client-side crash...

### ISSUE THREAD

**Issue by Korynkai** [Chicken-Bones/NotEnoughItems#155]
How's it going, Chicken-Bones?

I am currently working on a fork of MachineMuse's Powersuits and associated mods (Numina and MPSA), and, after finally eliminating an issue with overall JSON recipe handling, I seem to have come across another (apparently old) bug in Numina which results in this crash: https://gist.github.com/Korynkai/7ae60520ec8c7dfba989 when pressing the "U" key. It always happens when Numina and NEI are installed together.

Note: This isn't a high-priority issue because it is client-side only. However it is still a crash, elevating its initial priority anyway.

I was wondering, as developer of NEI, if you had any insight into the matter. As current maintainer of Numina (noting MachineMuse's amusing response in this issue: https://github.com/MachineMuse/MachineMusePowersuits/issues/564), I find this error message tells me nothing except that the crash starts at NEIServerUtils.java line 344, prints every line within NEI and Minecraft which is involved, even Sun/Java libraries and MultiMC, tells me it's an object that's trying to cast as ItemStack, ItemStack[] or List<ItemStack>, but completely misses Numina and anything associated with it.

Somehow I feel as if it's probably something silly and stupid I am missing, however, sometimes we all require a second set of eyes to find these "silly and stupid" problems, and yet I do feel as if this is going to require a collaborative effort to solve anyway.

Our GitLab with the current source for Numina is located here: https://git.qmxtech.com/open-source/numina/tree/master, the specific file in question is probably this file: https://git.qmxtech.com/open-source/numina/blob/master/src/main/scala/net/machinemuse/numina/nei/JSONRecipeHandler.java.
Do not use our GitHub for this, it is only a release mirror and has a server-side JSON bug which I have fixed in our internal build.

Thanks in advance! I hope we can track down and eliminate this old Numina + NEI issue!

_Forgot to note in this post: This is specifically for 1.7.10. Numina (and consequently Powersuits and MPSA) is not being developed for 1.8 as of yet._

**Comment by Korynkai** [comment]
I also may add that, while there is a debug string currently being printed each time each method in this class is called, none of these strings are actually being printed in the case of using the usage ('U' key) functionality of NEI, this is not the case when using the recipe ('R' key) functionality, which seems to work perfectly fine.

**Comment by Brycey92** [comment]
A lot of mods used to cause crashes when the U key was pressed over an item with NEI installed back in MC 1.5-1.6 as well. I think my modpack had the U key working for all of 1 week before something else broke it again. Perhaps there is a way to catch this in NEI as well instead of trying to fix it in each and every mod that creates such a crash.

**Comment by Korynkai** [comment]
Seems to me it would be preferred if this were caught in NEI in the first place with at least an error message pointing to the hook that made it crash... 
It seems NEI uses Java Objects as arguments in some places though, a practice similar to what is known in C++ as 'void pointers'. This is a _very_ bad practice which should be avoided at all costs unless you know _exactly_ what type(s) an object would be when it is passed to such a method, and is only thought of as decent practice in private or protected methods, where another method in the same class or a closely related class in the same project is calling this method. This is where method overloading comes into play though... For instance, you can define a function:
>void foo(int x, bool y)

and then define another function named exactly the same, perhaps it does something very similar but in a different manner, for instance, even calling the above method while setting the second argument (a boolean) to `true`:
>void foo(int x)

even:

>void foo(int x) {
&nbsp;&nbsp;&nbsp;&nbsp;foo(x, true);
}

There would be absolutely no errors in this case, as long as the signatures are different (you can't define two methods that have the same argument types and return the same type)...

Unfortunately I'm not sure if this is even what's going on. Actually, I think a null is being passed somewhere which isn't being checked before use, another bad practice (and something I heard @Chicken-Bones may be slightly infamous for? I may be thinking of another mod developer though...)

**Comment by Brycey92** [comment]
I couldn't have said it better myself, and I wholeheartedly agree with you; @Chicken-Bones has made some of the most essential mods around.

**Comment by Chicken-Bones** [comment]
Sorry it has taken me so long to reply to this.

The crash at NEIServerUtils.java:344 can be summarised as follows.
the forge class `ShapedOreRecipe`, has an `Object[] input` field, for storing the recipe inputs for each slot. 
In forge, and that class, the only valid contents are `ItemStack` or `List<ItemStack>` (see ShapedOreRecipe.java:211)
NEI has found an instance of ShapedOreRecipe that puts something different into the input field, (most likely an overriding class) and NEI doesn't know how to handle this.

I will make NEI simply discard handling of recipes that do this.

Note that it is NEI's `ShapedRecipeHandler` on the normal recipe list that is crashing, and that it only attempts to parse `ShapedRecipes` or `ShapedOreRecipe` instances, so it is not your `JSONRecipe` or handler that is the issue.

Now, Re: your short, nicely presented essay
When writing this piece of code, I added an instanceof chain, and I added an exception at the end, as is common practice, so that if there was something strange in that field at a later date, I would at least know that's what happened, and not let other things fall through and produce untraceable results.

It would be nice if NEI handled it, and just discarded handling the recipe, which is what I'll now make it do, but at the time, I didn't anticipate this scenario.

Listing the mod responsible is not possible in all cases, though printing the recipe class, and class of offending item would be useful information (which I'll add to the debug error message)

NEI and a few of my mods probably do use Java Objects in a few places, but they are only used internally, and all their potential values are known and constrained.

I am familiar with the reasons for plain Object fields being bad practice, and familiar with overloading, inheritance, interfaces and other aspects of the Java language, and attempt to keep my use of weak typing to a minimum.

You may have heard me as slightly infamous for not checking nulls. I don't tend to check nulls when something shouldn't be null, especially when I wouldn't know what to do with it if it was.
Consider this (contrived and simplistic) example, forge adds a tank interface, and I start filling the tank, then request some additional information for my operation, but it returns null somewhere, there is no documentation on null being a valid return value, and everything I've interfaced with returns good information here, in fact it's called somewhere else in forge with no null check, so I just go ahead and try and use it.
If I added a check, and it was null, well my operation is half complete, do I undo it, do I check for nulls beforehand on everything? Even if I did, the reason for a null return could be something completely unanticipated. So I could add an exception, but an NPE is an exception anyway, so I'd rather just wait for the NPE to appear, find out the cause, and fix it at the source, rather than attempting to anticipate scenarios that probably shouldn't exist.
Unfortunately this gets me caught a few times, but especially in a modding community, where no-one really knows what's expected of each other, things like this are bound to happen.

tldr; I don't null check as much as otheres, because I consider the small decrease in mod compatiblity issues to not be worth the additional effort, code convolution and potential problems. 

As for Optional, I'm aware of these, particularly Scala Option, and have used them in some cases. However, most of my null crashes are other mods interacting in some way that shouldn't return null but does, and is not an optional return. 
In most cases, I prefer to use null to Option, because of the reduced overhead. If part of an external API I make sure to document the use, validity and meaning of null in for that system.

In closing I'd like to thank you for your consideration and time, I'll push a fix soon.

**Comment by Korynkai** [comment]
Thank you, Chicken-Bones.  I would've gotten back sooner if I wasn't trying to get MPS 0.11.0.6-QMX out and I do completely see your point. I must say that as a programmer I respect you and I did realize that perhaps these situations are more likely to occur regardless, specifically with mods like NEI.. Anyway, keep up the good work and I look forward to the fix! ^_^

**Comment by Chicken-Bones** [comment]
My apologies for taking so long. Should be fixed as of d02fdc2f0bde3437c508ce9516e2a735fa24690d (1.7.10-1.0.4.89)

**Comment by Korynkai** [comment]
Awesome, thank you for your hard work! :smiley:
